src/Parser/Import.ml

```diff
@@ -0,0 +1,177 @@
+    if Sys.file_exists fname then Some fname else None)
+
+(** Convert a module path to a (relative) file path to be found within the
+    file system. *)
+let to_file_path (p, n) =
+  List.fold_right Filename.concat p (n ^ Config.src_extension)
+
+(** Find the file system path for a given absolute module path. *)
+let find_mod_abs (p, n) =
+  match p with
+  | pref :: p when pref = Config.local_mod_prefix ->
+    find_file_in !Config.local_search_dirs (to_file_path (p, n))
+  | [] | _ :: _ ->
+    find_file_in !Config.lib_search_dirs (to_file_path (p, n))
+
+(** Find a module relative to the module directory path [prefix], and return
+    the absolute module path and file system path. *)
+let find_mod_rel prefix p =
+  let rec loop prefix_r (p, n) =
+    let abs_path = (List.rev_append prefix_r p, n) in
+    match find_mod_abs abs_path with
+    | Some fname -> Some (abs_path, fname)
+    | None       ->
+      begin match prefix_r with
+      | []            -> None
+      | _ :: prefix_r -> loop prefix_r (p, n)
+      end
+  in
+  loop (List.rev prefix) p
+
+(** Find the absolute module path and file system path for an import path. *)
+let find_mod prefix (p : Raw.import_path) =
+  match p with
+  | IPAbsolute(p, n) -> find_mod_abs (p, n) |> Option.map (fun f -> (p, n), f)
+  | IPRelative(p, n) -> find_mod_rel prefix (p, n)
+
+(** Find and parse imported module relative to the path [prefix] and add it
+    to the [mods] map with the absolute mod identifier as the key.
+    Previously imported modules can be passed in the [imported] parameter
+    to avoid including them twice. *)
+let rec parse_import ~imported prefix mods (import : Raw.import) =
+  let path, new_name =
+    match import.data with
+    | IImportAs(path, new_name) -> (path, Some new_name)
+    | IImportOpen(path)         -> (path, None)
+  in
+  let path, fname =
+    match find_mod prefix path with
+    | Some path_fname -> path_fname
+    | None            ->
+      Error.fatal (Error.cannot_find_module import.pos (string_of_path path))
+  in
+  let name = internal_mod_id path in
+  let import = { import with data = (name, new_name) } in
+  if StrSet.mem name imported || StrMap.mem name mods then (mods, import)
+  else
+    let new_prefix = fst path in
+    let imports, defs = File.parse_defs ~pos:import.pos fname in
+    let mods = StrMap.add name ([], defs) mods in
+    let mods, imports =
+      parse_imports ~imported new_prefix mods imports in
+    (StrMap.add name (imports, defs) mods, import)
+
+and parse_imports ~imported prefix =
+  List.fold_left_map (parse_import ~imported prefix)
+
+(** Build a graph of all (transitively) imported modules. *)
+let collect_imports ~imported =
+  parse_imports ~imported [ Config.local_mod_prefix ] StrMap.empty
+
+let rec sort_dfs mods name on_path (visited, sorted) =
+  (* TODO: Once recursive module files are implemented cycle detection in this
+     function will likely become redundant. *)
```

Comment:
It depends on how we want to mark recursive module files. But I agree, we will change something here.